ADC stream API

vladislav-pejic opened the pull request:
Addition of streaming APIs for ADC devices that enables them to stream data to the application in more efficient way then it is currently possible.

Our goal was to provide a way to continuously stream data from ADC. We used similar approach as sensor stream API implementation.

On the user side adc_steam and adc_get_decoder APIs are added while on the driver side adc_driver_api is extended with submit and get_decoder.
Following decoder APIs are also added: get_frame_count, get_size_info, decode and has_trigger. Result of decode API should be in V in q31_t format.
Supported triggers: 
- ADC_TRIG_DATA_READY
- ADC_TRIG_FIFO_FULL
- ADC_TRIG_FIFO_WATERMARK
There are 3 supported operations on data associated with trigger:
- ADC_STREAM_DATA_INCLUDE - Include whatever data is associated with the trigger
- ADC_STREAM_DATA_NOP - Do nothing with the associated trigger data, it may be consumed later
- ADC_STREAM_DATA_DROP - Flush/clear whatever data is associated with the trigger

In this PR there are two driver implementations that are demonstrating use of streaming APIs with RTIO.

Signed-off-by: Vladislav Pejic vladislav.pejic@orioninc.com

dimitrije-lilic committed:
drivers: adc: max32: Support for RTIO stream

Updated MAX32 driver with RTIO  stream functionality.

Signed-off-by: Dimitrije Lilic <dimitrije.lilic@orioninc.com>

vladislav-pejic committed:
driver: adc: Added stream APIs for ADC

Introduce a streaming APIs for ADC devices.
Two new APIs are added to the adc_driver_api: submit and get_decoder.
Added decoder following APIs: get_frame_count, get_size_info, decode,
has_trigger.

Supported triggers are:
- ADC_TRIG_DATA_READY
- ADC_TRIG_FIFO_WATERMARK
- ADC_TRIG_FIFO_FULL
Supported operations to be done on trigger:
- include - whatever data is associated with the trigger
- nop - do nothing with data associated with the trigger
- drop - clear data associated with the trigger

Some changes to the linker scripts were needed to add decoder APIs.

Signed-off-by: Vladislav Pejic <vladislav.pejic@orioninc.com>

vladislav-pejic committed:
driver: adc: ad405x: Add RTIO stream

Added support for RTIO stream.
Also added sampling_period property to the DTS. This is for setting the
sampling period when streaming is used.
Hardware or kernel timer can be
used for this.
To use hardware timer you need to add it to the DTS. Example:

{
	chosen {
		zephyr,adc-clock = &counter0;
	};
};

Signed-off-by: Vladislav Pejic <vladislav.pejic@orioninc.com>

bjarki-andreasen commented:
The similarities between this `adc.h` driver extension and the existing sensing subsystem makes it hard for me to justify the complexity of the addition in this PR compared to simply adding a new data type and sensor type to the sensing API. Has it been considered to use the sensing subsystem to stream ADC data?

teburd reviewed (dismissed):
I read through this fairly carefully, primarily the the flow from timer/counter to the gpio coming back leading to a bus read. This all makes sense.

There's some questions I'd have here around efficiency and maybe there's ways of improving it that are worth exploring, but the flow all looks correct to me.

I have no strong opinion on the re-use or not of the existing ADC API and such here.

vladislav-pejic commented on drivers/adc/default_rtio_adc.c:
```diff
@@ -0,0 +1,323 @@
+/*
+ * Copyright (c) 2023 Google LLC.
+ * Copyright (c) 2024 Croxel Inc.
```
Corrected. Thank you.

vladislav-pejic commented:
> The similarities between this `adc.h` driver extension and the existing sensing subsystem makes it hard for me to justify the complexity of the addition in this PR compared to simply reusing `struct sensing_sensor_value_q31,` and adding a new sensor type `#define SENSING_SENSOR_TYPE_ANALOG_VOLTAGE` to the sensing API. Has it been considered to use the sensing subsystem to stream ADC data?

Could you please explain this in more detail? Did you mean to use `struct sensing_sensor_value_q31` instead of `struct adc_data`? I have never used sensing subsystem, so I did not consider using it for ADC stream. My intension was to create something similar to the sensor_stream.

vladislav-pejic commented on drivers/adc/adc_ad405x.c:
```diff
@@ -459,7 +569,7 @@ static int adc_ad405x_validate_buffer_size(const struct device *dev,
 }
 
-int ad405x_conv_start(const struct device *dev)
+static int ad405x_conv_start(const struct device *dev)
```
I didn't fully understand you here. Can you give me more details? Thank you.

ubieda replied:
This means using the counter API (as it is) to pace a periodic stream of conversions is likely to cause jitter and other kinds of issues. I think the existing `interval_us` parameter was only meant to cover successive one-shot readings, and not a steady stream of conversions:

https://github.com/zephyrproject-rtos/zephyr/blob/d8b37753061a9e0ace95204eccddee1f4b4c477b/include/zephyr/drivers/adc.h#L576-L585

bjarki-andreasen commented:
> > The similarities between this `adc.h` driver extension and the existing sensing subsystem makes it hard for me to justify the complexity of the addition in this PR compared to simply reusing `struct sensing_sensor_value_q31,` and adding a new sensor type `#define SENSING_SENSOR_TYPE_ANALOG_VOLTAGE` to the sensing API. Has it been considered to use the sensing subsystem to stream ADC data?
> 
> Could you please explain this in more detail? Did you mean to use `struct sensing_sensor_value_q31` instead of `struct adc_data`? I have never used sensing subsystem, so I did not consider using it for ADC stream. My intension was to create something similar to the sensor_stream.

ah, so, I notice that https://github.com/zephyrproject-rtos/zephyr/blob/b1f9351930f14775e90ebfb2d3d3998b6552ca1f/include/zephyr/sensing/sensing_datatypes.h#L67-L90 is quite similar to
https://github.com/zephyrproject-rtos/zephyr/blob/b1f9351930f14775e90ebfb2d3d3998b6552ca1f/include/zephyr/drivers/sensor_data_types.h#L52-L67 and even similar to 
```
struct adc_data {
	struct adc_data_header header;
	int8_t shift;
	struct adc_sample_data {
		uint32_t timestamp_delta;
		union {
			q31_t value;
		};
	} readings[1];
};
```
It seems like these are intended to be used as part of higher level subsystem, the sensing subsystem, so maybe there is some duplicate work that can be avoided? @huhebo @lixuzha

ttmut commented on drivers/adc/adc_max32.c:
```diff
@@ -276,6 +525,73 @@ static int adc_max32_init(const struct device *dev)
+
+	MXC_ADC_Handler();
+	if (flags & int_req) {
+		MXC_ADC_Free();
```
This will fail to build for some MAX32 SoCs, i.e. the ones with RevA ADCs.

vladislav-pejic replied:
Protection is added to limit support to only MAX32690 which uses RevB.

henrikbrixandersen commented:
> @teburd Can we put this on the agenda for the next Sensor WG meeting on June 2?

Why the sensor WG? This sounds more like a topic for the Arch WG as it is not directly sensor related? /CC @carlescufi

yperess commented on drivers/adc/adc_max32.c:
```diff
@@ -150,6 +209,196 @@ static int adc_max32_read(const struct device *dev, const struct adc_sequence *s
+		memcpy(out, buffer, 2);
+
+		buffer = sample_end;
+		*fit = (uintptr_t)sample_end;
```
I would strongly suggest not using memory addresses for the frame iterator. Would data offset work for you? `fit` should be 0 on a first read and you can increment it by the number of bytes that you read in the while loop.

yperess commented on drivers/adc/adc_max32.c:
```diff
@@ -276,6 +525,73 @@ static int adc_max32_init(const struct device *dev)
+
+	if (flags & WRAP_MXC_F_ADC_CONV_DONE_IF) {
+
+		data->timestamp = k_ticks_to_ns_floor64(k_uptime_ticks());
```
Should this use the sensor clock API?

vladislav-pejic committed:
boards: shields: EVAL-AD4052-ARDZ comp and overlay

Fix for compatible property.
Addition of hardware timer to be used with APARD32690 board.

Signed-off-by: Vladislav Pejic <vladislav.pejic@orioninc.com>

vladislav-pejic committed:
boards: adi: apard32690: Fix for gpio-map

Fix for gpio-map-mask and gpio-map-pass-thru to enable use of
MAX32_GPIO_VSEL_VDDIOH and other defines with arduino_header connector.

Signed-off-by: Vladislav Pejic <vladislav.pejic@orioninc.com>

vladislav-pejic committed:
sample: adc_stream: Add sample app for ADC stream

Add sample application to demonstrate new ADC stream APIs.
Two tests are also added:
- ad4052-stream
- max32-stream

Signed-off-by: Vladislav Pejic <vladislav.pejic@orioninc.com>

vladislav-pejic commented on drivers/adc/adc_ad405x.c:
```diff
@@ -899,17 +1015,327 @@ static int adc_ad405x_init(const struct device *dev)
+
+	if (data->data_ready_gpio > AD405X_GP1)	{
+		LOG_ERR("DATA_READY irq is not enabled!");
+		return;
```
Yes. I have added error reporting. Thank you.

vladislav-pejic commented on drivers/adc/adc_ad405x.c:
```diff
@@ -899,17 +1015,327 @@ static int adc_ad405x_init(const struct device *dev)
+	}
+
+	if (rc < 0) {
+		return;
```
Yes. I have added error reporting. Thank you.

vladislav-pejic commented on drivers/adc/adc_ad405x.c:
```diff
@@ -899,17 +1015,327 @@ static int adc_ad405x_init(const struct device *dev)
+		} else {
+			break;
+		}
+	}
```
I agree. I have removed entire function and replaced it with '32 - __builtin_clz(vref_mv)' because __builtin_clz considers input value as unsigned int (32 bit) and will return value according to that.

vladislav-pejic commented on drivers/adc/adc_ad405x.c:
```diff
@@ -899,17 +1015,327 @@ static int adc_ad405x_init(const struct device *dev)
+		return;
+	}
+
+	if (read_config->is_streaming == false) {
```
It was just another protection. I do not know scenario where is_streaming is true and sqe should be NULL. I have removed is_streaming check.

vladislav-pejic commented on drivers/adc/adc_max32.c:
```diff
@@ -150,6 +209,196 @@ static int adc_max32_read(const struct device *dev, const struct adc_sequence *s
+		} else {
+			break;
+		}
+	}
```
I have removed entire function and replaced it with '32 - __builtin_clz(vref_mv)' so we do not need to put it in the header.

vladislav-pejic commented on drivers/adc/adc_ad405x.c:
```diff
@@ -899,17 +1015,327 @@ static int adc_ad405x_init(const struct device *dev)
+		return;
+	}
+
+	data->timestamp = (uint64_t)(counter_ticks_to_us(data->timer_dev, ticks))
+				* 1000 /* uS to nS */;
+#else
+	data->timestamp = k_ticks_to_ns_floor64(k_uptime_ticks());
```
I do not think it should. Sensor API will default to the system time if no counter/timer is defined. This will be basically the same just without layers of abstraction.

vladislav-pejic commented on drivers/adc/Kconfig.ad405x:
```diff
@@ -11,6 +11,14 @@ config ADC_AD405X
+config AD405X_STREAM
+	bool "Use FIFO to stream data"
+	select AD405X_TRIGGER
+	default n
```
Removed

vladislav-pejic commented on drivers/adc/Kconfig.max32:
```diff
@@ -10,3 +10,10 @@ config ADC_MAX32
+config ADC_MAX32_STREAM
+	bool "Use FIFO to stream data"
+	select ADC_ASYNC
+	default n
```
Removed

vladislav-pejic commented on drivers/adc/adc_ad405x.c:
```diff
@@ -174,9 +179,94 @@ struct adc_ad405x_data {
+static void ad405x_stream_irq_handler(const struct device *dev);
+static int ad405x_conv_start(const struct device *dev);
+
+#if DT_HAS_CHOSEN(zephyr_adc_clock)
```
Maybe we do not but for me it is easier to see what is used and when.

ubieda commented on drivers/adc/adc_ad405x.c:
```diff
@@ -899,17 +1015,322 @@ static int adc_ad405x_init(const struct device *dev)
 }
 
+#ifdef CONFIG_AD405X_STREAM
+void ad405x_submit_stream(const struct device *dev, struct rtio_iodev_sqe *iodev_sqe)
```
Reinforcing the jitter concern above: If this works the same way the Sensor Streaming works, this is a multi-shot submission; meaning: as soon as the previous conversion finishes, this is called again. As such, we shouldn't start a timer (which then restarts a conversion) in here.

Suggestion: Figure offloading the timer pacing to a periodic timer (ideally, hardware timer which enables the ADC conversion), and only restart it if the config-set changes. Some examples of this existing pattern:
https://github.com/zephyrproject-rtos/zephyr/blob/ed7f2e0833c25fdc67e0d2c1e6d6915275407569/drivers/sensor/tdk/icm45686/icm45686_stream.c#L476-L492

vladislav-pejic replied:
I agree that this is a valid concern but IMO this has more to do with driver then the stream API. Since there is no way to set this up with use of counter APIs can leave this for the future as a driver improvement?

ubieda commented on drivers/adc/adc_ad405x.c:
```diff
@@ -174,9 +179,94 @@ struct adc_ad405x_data {
+
+	counter_set_channel_alarm(data->timer_dev, 0, &alarm_cfg);
+#else
+	k_timer_start(&data->sample_timer, K_USEC(cfg_405->sampling_period), K_NO_WAIT);
```
IMO - This API is missing a way to configure the streaming mode. For instance: setting the watermark level, or the DRDY period (like this).

A default DTS property shouldn't be the only way we'd set these things.

teburd replied:
This is definitely something that could/should be looked at as configuration, the adc configuration today mostly comes int he adc_sequence struct (interval_us, channels enabled, resolution, oversampling, etc)

Arguably adc_sequence could be changed slightly to extract all the configuration away from the buffer pointer and len pair that are in it to make this a bit cleaner.

That would likely require updating many drivers. Please file an issue to make this a requirement for stabilizing this ADC stream API and I'd be ok with the existing setup.

ubieda reviewed (commented):
There's quite a bit to digest in this PR, so I mostly focused on the API and how it's used. Not so much on the details in the drivers.

In principle I like this API and how it can help set up an ADC data-stream more efficiently. There's some concerns on configurability which I believe can be served with an additional API.

Hopefully this feedback helps to get the conversation forward.

MaureenHelm reviewed (commented):
> > > @teburd Can we put this on the agenda for the next Sensor WG meeting on June 2?
> > 
> > 
> > Why the sensor WG? This sounds more like a topic for the Arch WG as it is not directly sensor related? /CC @carlescufi
> 
> It's based on the sensor streaming API and I wanted to discuss in a smaller forum with people familiar with that API before bringing to the Arch WG.

Notes from the Sensor WG discussion:
* Consensus: Doesn't make sense to combine this with sensors
* Should use iodev for device specific/streaming specific config
* Needs a sample and should be shown at Arch WG
* Does seem to, at an API level, solve the continuous streaming of ADCs which is great and has been long sought after

https://lists.zephyrproject.org/g/sensors-wg/message/146

@carlescufi can we put this on the agenda for an upcoming Arch WG meeting? Tomorrow is too soon, maybe next week?

vladislav-pejic commented on drivers/adc/adc_ad405x.c:
It would be more readable to separate regular driver and stream functionality into separate files but folder structure for ADC drivers currently does not support this. For each driver there is only one .c file and every feature is implemented there. Also, there is no (or at least they are rear) header files for specific ADC drivers. Changing all of that will be a lot and maybe we should leave it for some future PR?

bjarki-andreasen reviewed (commented):
A few comment regarding the sample, it will be crucial for users to understand how to adopt adc_stream, so add more comments, mention stuff twice, make sure nothing is implied as much as possible :)

vladislav-pejic commented:
> A few comment regarding the sample, it will be crucial for users to understand how to adopt adc_stream, so add more comments, mention stuff twice, make sure nothing is implied as much as possible :)

Thank you for the feedback. I added comments and corrected everything that you have suggested. Can you check if the changes are ok?

bjarki-andreasen commented:
> > A few comment regarding the sample, it will be crucial for users to understand how to adopt adc_stream, so add more comments, mention stuff twice, make sure nothing is implied as much as possible :)
> 
> Thank you for the feedback. I added comments and corrected everything that you have suggested. Can you check if the changes are ok?

changes look good :)

MaureenHelm commented on drivers/adc/adc_max32.c:
```diff
@@ -150,6 +150,163 @@ static int adc_max32_read(const struct device *dev, const struct adc_sequence *s
 }
 
+<<<<<<< Updated upstream
+=======
```
Oops

MaureenHelm commented on drivers/adc/adc_max32.c:
```diff
@@ -150,6 +150,163 @@ static int adc_max32_read(const struct device *dev, const struct adc_sequence *s
+}
+#endif /* CONFIG_ADC_MAX32_STREAM */
+
+>>>>>>> Stashed changes
```
Oops

bjarki-andreasen commented on drivers/adc/default_rtio_adc.c:
```diff
@@ -0,0 +1,330 @@
+	const struct device *dev = cfg->adc;
+	const struct adc_driver_api *api = dev->api;
+
+	if (api->submit != NULL) {
+		api->submit(dev, iodev_sqe);
+	} else if (!cfg->is_streaming) {
+		adc_submit_fallback(dev, iodev_sqe);
+	} else {
+		rtio_iodev_sqe_err(iodev_sqe, -ENOTSUP);
+	}
```
This pattern will force the inclusion of the default backend, and thus RTIO_WORKQ, which I believe will unconditionally add a thread pool if ADC_STREAM is enabled, regardless of whether its used or not. If this is indeed the case, the fallback should be optional, and not conflated with ADC_STREAM

vladislav-pejic replied:
The idea with this fallback mechanism was when we add tests for ADC_STREAM (similarly to the tests for SENSOR_STREAM), ADC devices that don't have streaming functionality can pass them. Default backend was never tested and this part I was planning to revisit at a later date and as a separate PR together with the test. Is that acceptable for you?

teburd commented:
Is @anangl around to review this as maintainer? It's been several months.

nordicjm commented on drivers/adc/Kconfig.ad405x:
```diff
@@ -11,6 +11,13 @@ config ADC_AD405X
+config AD405X_STREAM
+	bool "Use FIFO to stream data"
+	select AD405X_TRIGGER
+	depends on SPI_RTIO
```
`depends on ADC_AD405X` I had to fix the below bleeding Kconfig the other day, let's not add more

nordicjm commented on drivers/adc/Kconfig.max32:
```diff
@@ -10,3 +10,16 @@ config ADC_MAX32
 	help
 	  Enable ADC driver for ADI MAX32xxx MCUs.
+
+config ADC_MAX32_REVB_ME18
+	bool "revb"
+	help
+	  Use this configuration option when SoC supports revb and me18.
+	  This configuration option is necessary if ADC_STREAM is used.
```
this again should be in dts, why is it in kconfig? Comment has not been addressed: https://github.com/zephyrproject-rtos/zephyr/pull/90285#discussion_r2340896384

vladislav-pejic replied:
Hi @nordicjm. Can you give me more information or some example how you expect this to be implemented? Thank you.

nordicjm commented on drivers/adc/Kconfig.max32:
```diff
@@ -10,3 +10,16 @@ config ADC_MAX32
+config ADC_MAX32_STREAM
+	bool "Use FIFO to stream data"
+	select ADC_ASYNC
+	depends on ADC_MAX32_REVB_ME18
```
`depends on ADC_MAX32`
